dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/writer/GeneWriter.java

```diff
@@ -17,139 +17,478 @@
 import com.fasterxml.jackson.databind.JsonNode;
-import com.fasterxml.jackson.databind.MappingIterator;
 import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.node.ArrayNode;
 import com.fasterxml.jackson.databind.node.ObjectNode;
-import com.google.common.base.Stopwatch;
 import com.mongodb.MongoClientURI;
 
-import de.undercouch.bson4jackson.BsonFactory;
+import lombok.NonNull;
+import lombok.SneakyThrows;
+import lombok.val;
+import lombok.extern.slf4j.Slf4j;
 
-/**
- * Imports from Heliotrope {@code genes.bson} {@code mongodump} file into DCC gene database.
- */
 @Slf4j
-public class GeneWriter extends AbstractJongoWriter<InputStream> {
+public class GeneWriter extends AbstractJongoWriter<ObjectNode> {
 
   /**
-   * How frequently to report status on inserted genes.
+   * Constants
    */
   private static final int STATUS_GENE_COUNT = 10000;
+  private static final ObjectMapper MAPPER = new ObjectMapper();
+  private static final Pattern TSV = Pattern.compile("\t");
+
+  /**
+   * Dependencies
+   */
+  private final BufferedReader bufferedReader;
+  private final Map<String, String> summaryMap;
+  private final Map<String, String> nameMap;
+  private final Map<String, ArrayNode> synMap;
+  private final Map<String, String> canonicalMap;
+  private final Map<String, List<ProteinFeature>> pFeatures;
+  private final Map<String, ObjectNode> externalIds;
+  private int counter = 0;
+  private MongoCollection geneCollection;
 
-  public GeneWriter(MongoClientURI mongoUri) {
+  public GeneWriter(@NonNull MongoClientURI mongoUri, @NonNull BufferedReader bufferedReader,
+      @NonNull Map<String, String> summaryMap, @NonNull Map<String, String> nameMap,
+      @NonNull Map<String, ArrayNode> synMap, @NonNull Map<String, String> canonicalMap,
+      @NonNull Map<String, List<ProteinFeature>> pFeatures, @NonNull Map<String, ObjectNode> externalIds) {
     super(mongoUri);
+    this.bufferedReader = bufferedReader;
+    this.geneCollection = getCollection(ReleaseCollection.GENE_COLLECTION);
+    this.geneCollection.drop();
+    this.summaryMap = summaryMap;
+    this.nameMap = nameMap;
+    this.canonicalMap = canonicalMap;
+    this.synMap = synMap;
+    this.pFeatures = pFeatures;
+    this.externalIds = externalIds;
+  }
+
+  /**
+   * Method responsible for constructing the skeleton of the gene model. It streams the gtf file, constructing the
+   * elements of the gene model, and joining in information from the maps provided to the constructor of the class.
+   * Write Gene to mongo as soon as it is constructed and moves onto next one.
+   */
+  @SneakyThrows
+  public void consumeGenes() {
+    log.info("CONSUMING GENES");
+    ObjectNode geneNode = null;
+    ObjectNode curTranscript = null;
+    ArrayNode transcripts = MAPPER.createArrayNode();
+    ArrayNode exons = MAPPER.createArrayNode();
+
+    for (String s = bufferedReader.readLine(); null != s; s = bufferedReader.readLine()) {
```

Comment:
This needs to be broken down like so much Jack and the Beanstalk. Please create some helper methods.